README_tcpdevmem.md

```diff
@@ -0,0 +1,277 @@
+  --client \
+  --mode cuda
+```
+
+
+
+### Building your own image for testing
+
+
+```
+# within the repo directory
+docker build -t neperc .
+
+function run_neper_container() {
+  docker run \
+    --name neper_c \
+    -u 0 --network=host \
+    --cap-add=IPC_LOCK \
+    --userns=host \
+    --volume /run/tcpx:/tmp \
+    --volume /var/lib/nvidia/lib64:/usr/local/nvidia/lib64 \
+    --volume /var/lib/tcpx:/usr/local/tcpx \
+    --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 \
+    --device /dev/nvidia0:/dev/nvidia0 \
+    --device /dev/nvidia1:/dev/nvidia1 \
+    --device /dev/nvidia2:/dev/nvidia2 \
+    --device /dev/nvidia3:/dev/nvidia3 \
+    --device /dev/nvidia4:/dev/nvidia4 \
+    --device /dev/nvidia5:/dev/nvidia5 \
+    --device /dev/nvidia6:/dev/nvidia6 \
+    --device /dev/nvidia7:/dev/nvidia7 \
+    --device /dev/nvidia-uvm:/dev/nvidia-uvm \
+    --device /dev/nvidiactl:/dev/nvidiactl \
+    --cap-add=NET_ADMIN \
+    --env LD_LIBRARY_PATH=/usr/local/nvidia/lib64:/usr/local/tcpx/lib64 \
+    "$@"
+}
+
+run_neper_container -it neperc bash
+```
+
+You can replace the `neperc` image name and run `docker push ${image_name}` if you’re building the image on one host, and running a container on another.
+
+
+### Override CUDA library directory (DLVM)
+
+The script assumes that `libcuda.so*` files are found in `/var/lib/nvidia/lib64`. In case this isn’t true (like when on DLVM), you can override the default env var: `CUDA_LIB_DIR`:
+
+```
+CUDA_LIB_DIR=/usr/lib/x86_64-linux-gnu ./run_neper_container.sh
+```
+
+
+## TCPDirect UDMA: Compiling tcp_stream
+
+Neper can be built statically on a host with UDMA header files.
+
+The Makefile assumes that the header files are found at `~/cos-kernel/usr/include`. You can manually change the directory in the Makefile:
+
+
+```
+CFLAGS += -DWITH_TCPDEVMEM_UDMA -DNDEBUG=1 -static -I ~/cos-kernel/usr/include <your new directory>
+```
+
+```
+# clone the Neper repository and checkout the tcpd branch
+git clone -b tcpd https://github.com/google/neper.git
+cd neper
+make tcp_steam WITH_TCPDEVMEM_UDMA=1
+
+# copy the binary to your hosts
+scp tcp_stream root@${HOST1}:~/
+scp multi_neper.py root@${HOST1}:~/
+
+scp tcp_stream root@${HOST2}:~/
+scp multi_neper.py root@${HOST2}:~/
+```
+
+
+
+## Running tcp_stream
+
+
+### Added flags
+
+In general, these flags will be automatically populated by `multi_neper.py`.
+
+```
+--tcpd-validate     # payload validation - must pass to both Tx/Rx if enabled
+--tcpd-tcpd-rx-cpy
+--tcpd-nic-pci-addr
+--tcpd-gpu-pci-addr
+--tcpd-phys-len     # CUDA mode allows for a much larger value than UDMA mode
+--tcpd-src-ip
+--tcpd-dst-ip
+--tcpd-link-name
+--queue-start
+--queue-num
+```
+
+Running TCPDirect requires the toggling of a handful of ethtool commands on the receiver (host). If running tcp_stream via `multi_neper.py`, this will automatically be done before each run.
+
+Otherwise, it might be necessary to run these commands before each tcp_stream run.
```

Comment:
I looked at what multi_neper.py does, and at the moment I don't think it's up to par. tcp_stream needs to be a reasonably standalone binary that works by itself, for devmem as it is for regular TCP. tcp_stream should not require a wrapper script to do machine set up for it for regular tcp or devmem.

Currently, multi_neper.py does:

1. Toggle header split. That is not correct. Header split toggeling should be done in neper after the dma-buf is bound and after the dma-buf is unbound. It's not correct to do a DMA_BUF_PAGES_BIND_RX and not do a header split toggle right after.
2. configure rss. This is reasonable I think.
3. Turn on ntuple (flow steering). since tcp_stream configures flow steering, I think it should also turn it on inside.
4. Tun on enable-max-rx-buffer-size: that is reasonable I think.

Please file a bug to move the header split & flow steering config fully inside tcp_stream as follow up work.